plutus-tx/src/PlutusTx/Builtins/Internal.hs

```diff
@@ -1068,3 +1071,18 @@ caseInteger i b = b !! fromIntegral i
 casePair :: forall a b r. BuiltinPair a b -> (a -> b -> r) -> r
 casePair p f = f (PlutusTx.Builtins.Internal.fst p) (PlutusTx.Builtins.Internal.snd p)
 {-# INLINE casePair #-}
+
+data CaseDataConstrBranch = forall a. CaseDataConstrBranch a
+
+caseDataConstr :: forall r. BuiltinData -> [CaseDataConstrBranch] -> r
+caseDataConstr d branches =
+  let
+    constr = unsafeDataAsConstr d
+    idx = PlutusTx.Builtins.Internal.fst constr
+    ds = PlutusTx.Builtins.Internal.snd constr
+    unwrap (CaseDataConstrBranch x) = unsafeCoerce x
+    applyToBranch :: r -> BuiltinList BuiltinData -> r
+    applyToBranch body =
+      caseList' body (\arg -> applyToBranch (unsafeCoerce body $ arg))
+  in applyToBranch (unwrap $ branches !! fromIntegral idx) ds
+{-# OPAQUE caseDataConstr #-}
```

Comment:
This is horrendous; needed that existential type since number of argument each branch handler takes is arbitrary. This works, but it's impossible to make it also work for when we disable builtin-casing compilation option. 

I'm going to try other ways.

Reply:
Made some changes. It works good now without this atrocity